lib/mayaUsd/nodes/proxyShapeBase.cpp

```diff
@@ -539,29 +584,56 @@ SdfLayerRefPtr MayaUsdProxyShapeBase::computeSessionLayer(MDataBlock&) { return
 
 #endif
 
-MStatus MayaUsdProxyShapeBase::computeInStageDataCached(MDataBlock& dataBlock)
+namespace {
+
+void remapSublayerRecursive(
+    const SdfLayerRefPtr&              layer,
+    std::map<std::string, std::string> remappedLayers)
 {
-    MStatus retValue = MS::kSuccess;
+    if (remappedLayers.empty())
+        return;
 
-    MDataHandle inDataHandle = dataBlock.inputValue(inStageDataAttr, &retValue);
-    CHECK_MSTATUS_AND_RETURN_IT(retValue);
+    bool                        updated = false;
+    SdfSubLayerProxy            sublayerPaths = layer->GetSubLayerPaths();
+    std::vector<SdfLayerRefPtr> sublayers;
+    std::vector<std::string>    newSublayerPaths;
+    newSublayerPaths.reserve(sublayerPaths.size());
+    for (auto path : sublayerPaths) {
+        if (remappedLayers.find(path) != remappedLayers.end()) {
+            updated = true;
+            if (!remappedLayers[path].empty())
+                newSublayerPaths.push_back(remappedLayers[path]);
+            remappedLayers.erase(path);
+        } else {
+            newSublayerPaths.push_back(path);
+        }
+        SdfLayerRefPtr subLayer = SdfLayer::Find(path);
+        if (subLayer) {
+            sublayers.push_back(subLayer);
+        }
+    }
+    if (updated) {
+        layer->SetSubLayerPaths(newSublayerPaths);
+    }
 
-    // If inData has an incoming connection, then use it. Otherwise generate stage from the filepath
-    if (!inDataHandle.data().isNull()) {
-        //
-        // Propagate inData -> inDataCached
-        //
-        MDataHandle inDataCachedHandle = dataBlock.outputValue(inStageDataCachedAttr, &retValue);
-        CHECK_MSTATUS_AND_RETURN_IT(retValue);
+    if (!remappedLayers.empty()) {
+        for (auto sublayer : sublayers) {
+            remapSublayerRecursive(sublayer, remappedLayers);
```

Comment:
During iteration, you can discover that all layers were remapped. So you can save both searching for layers, vector insert and calls to remapSublayerRecursive if you do something like:
```
if (remappedLayers.empty())
    return;
for(const auto& sublayerPath : sublayerPaths) {
    SdfLayerRefPtr subLayer = SdfLayer::Find(sublayerPath);
    remapSublayerRecursive(sublayer, remappedLayers);
    if (remappedLayers.empty())
        return;
}
```

Reply:
Done